sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala

```diff
@@ -30,8 +36,156 @@ case class GpuJsonToStructs(
+          }
+        }
+        closeOnExcept(isNullOrEmptyInput) { _ =>
+          withResource(isNullOrEmptyInput.ifElse(emptyRow, stripped)) { cleaned =>
+            withResource(cudf.Scalar.fromString("\n")) { lineSep =>
+              withResource(cudf.Scalar.fromString("\r")) { returnSep =>
+                withResource(cleaned.stringContains(lineSep)) { inputHas =>
+                  withResource(inputHas.any()) { anyLineSep =>
+                    if (anyLineSep.isValid && anyLineSep.getBoolean) {
+                      throw new IllegalArgumentException("We cannot currently support parsing " +
+                          "JSON that contains a line separator in it")
+                    }
+                  }
+                }
+                withResource(cleaned.stringContains(returnSep)) { inputHas =>
+                  withResource(inputHas.any()) { anyReturnSep =>
+                    if (anyReturnSep.isValid && anyReturnSep.getBoolean) {
+                      throw new IllegalArgumentException("We cannot currently support parsing " +
+                          "JSON that contains a carriage return in it")
+                    }
+                  }
+                }
+              }
+              (isNullOrEmptyInput, cleaned.joinStrings(lineSep, emptyRow))
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // Process a sequence of field names. If there are duplicated field names, we only keep the field
+  // name with the largest index in the sequence, for others, replace the field names with null.
+  // Example:
+  // Input = [("a", StringType), ("b", StringType), ("a", IntegerType)]
+  // Output = [(null, StringType), ("b", StringType), ("a", IntegerType)]
+  private def processFieldNames(names: Seq[(String, DataType)]): Seq[(String, DataType)] = {
+    val existingNames = Set[String]()
+    names.foldRight(Seq[(String, DataType)]())((elem, acc) => {
+      val (name, dtype) = elem
+      if (existingNames(name)) (null, dtype)+:acc else {existingNames += name; (name, dtype)+:acc}})
+  }
+
+  private def getSparkType(col: cudf.ColumnView): DataType = {
+    col.getType match {
+      case cudf.DType.INT8 | cudf.DType.UINT8 => ByteType
+      case cudf.DType.INT16 | cudf.DType.UINT16 => ShortType
+      case cudf.DType.INT32 | cudf.DType.UINT32 => IntegerType
+      case cudf.DType.INT64 | cudf.DType.UINT64 => LongType
+      case cudf.DType.FLOAT32 => FloatType
+      case cudf.DType.FLOAT64 => DoubleType
+      case cudf.DType.BOOL8 => BooleanType
+      case cudf.DType.STRING => StringType
+      case cudf.DType.LIST => ArrayType(getSparkType(col.getChildColumnView(0)))
+      case cudf.DType.STRUCT =>
+        val structFields = ArrayBuffer.empty[StructField]
+        (0 until col.getNumChildren).foreach { i =>
+          val child = col.getChildColumnView(i)
+          structFields += StructField("", getSparkType(child))
+        }
```

Comment:
I don't think we need a mutable here
```suggestion
        val structFields = (0 until col.getNumChildren).map { i =>
          val child = col.getChildColumnView(i)
          StructField("", getSparkType(child))
        }
```